### PR TITLE
Add support for multiple samples

### PR DESCRIPTION
**Pull request by jacobaustin123**
This is an initial draft PR for adding support for multiple samples. There is significant work to be done still, but the basic approach is to shift from any reliance on converting Variables to their value_t, and instead using `get_values(int i)` to access entries within the Variable. 

To do:
2. Add support for VariableView class that views but does not own external data.
3. Improve data storage in Variable, e.g. create an `allocator` that uses stack allocation whenever possible.

### CODE REVIEW

File: include/autoppl/expression/variable/constant.hpp [hunk 1]
```diff
@@ -11,9 +11,12 @@ struct Constant : util::VarExpr<Constant<ValueType>>
...
         : c_{c}
     {}
-    explicit operator value_t() const { return get_value(); }
-    value_t get_value() const { return c_; }
-
+    value_t get_value(int i) const {
+        assert(i >= 0);
+        return c_;
+    }
+    size_t size() const { return 1; }
```
Comment (JamesYang007) [inline include/autoppl/expression/variable/constant.hpp]: can be made `constexpr` :D

File: include/autoppl/algorithm/mh.hpp [hunk 2]
```diff
@@ -84,7 +84,7 @@ inline void mh_posterior__(ModelType& model,
             using state_t = typename util::var_traits<var_t>::state_t;
 
             if (var.get_state() == state_t::parameter) {
-                auto curr = var.get_value();
+                auto curr = var.get_value(0);
```
Comment (JamesYang007) [inline include/autoppl/algorithm/mh.hpp:86]: Just to be clear, this is because we're still assuming parameters are univariate and "one sample" right?

Comment (jacobaustin123) [reply include/autoppl/algorithm/mh.hpp:86]: Yes. For any parameter, the first vector element is used for the value. We could have a separate variable for the parameter value if that is more elegant.

File: include/autoppl/variable.hpp [hunk 3]
```diff
@@ -27,28 +30,46 @@ struct Variable : util::Var<Variable<ValueType>>
...
 
     Variable(pointer_t storage_ptr) noexcept
         : Variable(0, storage_ptr, state_t::parameter)
     {}
 
+    Variable(std::initializer_list<value_t> values) noexcept
+        : Variable(values.begin(), values.end(), nullptr, state_t::data) {}
+
+    Variable(value_t value, pointer_t storage_ptr, state_t state) noexcept
+        : storage_ptr_{storage_ptr}, state_{state} {
```
Comment (JamesYang007) [inline include/autoppl/variable.hpp]:
```cpp
: values_({value}), storage_ptr_{storage_ptr}...
```
may be more efficient depending on implementation.

File: test/algorithm/mh_unittest.cpp [hunk 4]
```diff
@@ -96,6 +97,22 @@ TEST_F(mh_fixture, sample_unif_normal_posterior_mean_stddev)
     EXPECT_NEAR(sample_average(storage_2), 1.868814361437099766, 0.1);
 }
 
+TEST_F(mh_fixture, sample_unif_normal_posterior_mean_stddev_samples) {
+    auto model = (
+        theta |= uniform(-1., 2.), 
+        y |= normal(theta, 1.0) // {0.1, 0.2, 0.3, 0.4, 0.5}
+    );
+
+    auto start = std::chrono::high_resolution_clock::now();
```
Comment (JamesYang007) [inline test/algorithm/mh_unittest.cpp]: oops this was from my temporary benchmarking. Do you mind removing this xD

File: test/expression/variable/variable_unittest.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    expected_size = 1;
+    real_size = var2.size();
+    
+    EXPECT_EQ(expected_size, real_size);
+
+    EXPECT_EQ(var1.get_value(0), 1.0);
+    EXPECT_EQ(var1.get_value(1), 2.0);
+    EXPECT_EQ(var1.get_value(2), 3.0);
+
+    EXPECT_DEATH({
```
Comment (JamesYang007) [inline test/expression/variable/variable_unittest.cpp]: Very cool 👍

**Review by JamesYang007 — COMMENTED** [review]
Looks great!!!!

**Review by JamesYang007 — CHANGES_REQUESTED** [review]
Just requesting that 1 change above ^

File: test/algorithm/mh_unittest.cpp (hunk 4 above)
Comment (jacobaustin123) [reply test/algorithm/mh_unittest.cpp]: Fixed in the latest version.

**Review by JamesYang007 — APPROVED** [review]
<3

File: include/autoppl/expression/model/eq_node.hpp [hunk 6]
```diff
@@ -44,17 +44,21 @@ struct EqNode : util::ModelExpr<EqNode<VarType, DistType>>
...
      * Compute log-pdf of underlying distribution with underlying value.
      * Assumes that underlying value has been assigned properly.
      */
-    dist_value_t log_pdf() const
-    { return dist_.log_pdf(orig_var_ref_.get().get_value()); }
+    dist_value_t log_pdf() const {
+        return dist_.log_pdf(get_variable());
+    }
 
+    auto& get_variable() const { return orig_var_ref_.get(); }
```
Comment (JamesYang007) [inline include/autoppl/expression/model/eq_node.hpp]: `const auto&` for extra clarity

**Review by JamesYang007 — APPROVED** [review]

**Review by JamesYang007 — CHANGES_REQUESTED** [review]
Sorry that one change above ^ but yeah it's perfect otherwise :')

**Review by JamesYang007 — APPROVED** [review]
Looks great!

**Pull request merged**